### PR TITLE
Add batchnorm folding transformations (fx)

### PR DESCRIPTION
**Pull request by fxmarty**
# What does this PR do?

Add the folding of `nn.BatchNorm2d` into `nn.Conv2d`, and of `nn.BatchNorm1d` into `nn.Linear`.

The former is useful for example for resnet, beit, regnet, dpt, levit, segformer and others.
The latter is useful solely for groupvit, and could be for levit; but there is a flatten inbetween so the current transformation is not directly usable: https://github.com/huggingface/transformers/blob/ab2006e3d6db88654526a4169e65d4bfc52da2e3/src/transformers/models/levit/modeling_levit.py#L153-L156

Fixes https://github.com/microsoft/onnxruntime/issues/12522

## Before submitting
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/fx/optimization/transformations.py [hunk 1]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+    Gain: 29.12 % in latency
+    ```
+
+    Note that the gain may depend on the batch size, image size, hardware used.
+    """
+
+    preserves_computation = True
+
+    def transform(self, graph_module: "GraphModule") -> "GraphModule":
+        modules = dict(graph_module.named_modules())
```
Comment (michaelbenayoun) [inline optimum/fx/optimization/transformations.py:643]: Same comment as above.

File: optimum/fx/optimization/transformations.py [hunk 2]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+        return fused_conv
+
+
+@add_end_docstrings(_ATTRIBUTES_DOCSTRING)
+class FuseBatchNorm1dInLinear(Transformation):
+    """
+    Transformation that fuses `nn.BatchNorm1d` following `nn.Linear` into a single `nn.Linear`.
+    The fusion will be done only if the linear layer has the batch normalization as sole following node.
+
+    Example on GPU:
```
Comment (michaelbenayoun) [inline optimum/fx/optimization/transformations.py:583]: Same comment as above.

File: optimum/fx/optimization/transformations.py [hunk 3]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+                    parent_name, _, name = batch_norm_node.target.rpartition(".")
+                    delattr(modules[parent_name], name)
+
+                    graph_module.graph.erase_node(node)  # delete reshape_as
+                    graph_module.graph.erase_node(batch_norm_node)  # delete BatchNorm1d
+                    graph_module.graph.erase_node(flatten_node)  # delete flatten
+        return graph_module
+
+    def fuse(self, linear: torch.nn.Linear, bn1d: torch.nn.BatchNorm1d):
+        fused_linear = copy.deepcopy(linear)
```
Comment (michaelbenayoun) [inline optimum/fx/optimization/transformations.py:738]: Same comment as the first fuse.

File: tests/fx/optimization/test_transformations.py [hunk 4]
```diff
@@ -211,59 +221,189 @@ def reverse(self, graph_module):
                 self.assertFalse(t1.transformed(node))
 
 
-def test_merge_linears():
-    _, traced = get_bert_model()
-    transformation = MergeLinears()
-    transformed = transformation(traced)
+class CustomTransformationsTests(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply tests/fx/optimization/test_transformations.py:224]: Otherwise the tests are not discovered in https://github.com/huggingface/optimum/blob/69f2883d945ccc669bdd0c78e0ba8fe4265c6792/.github/workflows/test_fx.yml#L37

By the way, I just moved `test_transformations.py` to `tests/fx`, otherwise is it not discovered either.

[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply tests/fx/optimization/test_transformations.py]: We should change that to:
```
python -m pytest tests/fx
```

That way we dont have to create unnecessary classes, and we can keep the directory tree.

File: optimum/fx/optimization/transformations.py [hunk 5]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+    def transform(self, graph_module: "GraphModule") -> "GraphModule":
+        modules = dict(graph_module.named_modules())
+
+        for node in graph_module.graph.nodes:
+            if node.op == "call_module" and node.args[0].op == "call_module":
+                if (
+                    type(modules[node.target]) is torch.nn.BatchNorm2d
+                    and type(modules[node.args[0].target]) is torch.nn.Conv2d
+                ):
+                    if len(node.args[0].users) > 1:  # Output of conv is used by other nodes
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/fx/optimization/transformations.py:540]: This means the output of the Conv2d goes to an other branch than BatchNorm2d, hence folding is not doable.
Like this:
```
    Conv2d
     /   \
    /     \
  ReLU   BatchNorm2d
```

Arguably we could support:
```
           Conv2d
     ______|     \
    /             \
BatchNorm2d      BatchNorm2d
```
but I don't think it's a commonn case.

File: optimum/fx/optimization/transformations.py [hunk 6]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+        bn_w = bn2d.weight if bn2d.weight is not None else torch.ones_like(bn2d.running_mean)
+        bn_b = bn2d.bias if bn2d.bias is not None else torch.ones_like(bn2d.running_mean)
+
+        bn_var_rsqrt = torch.rsqrt(bn2d.running_var + bn2d.eps)
+
+        fused_conv.weight = torch.nn.Parameter(
+            conv2d.weight * (bn_w * bn_var_rsqrt).reshape([-1] + [1] * (len(conv2d.weight.shape) - 1))
+        )
+
+        fused_conv.bias = torch.nn.Parameter(conv_b - bn2d.running_mean * bn_var_rsqrt * bn_w + bn_b)
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/fx/optimization/transformations.py:572]: I did not manage to do it inplace in the case where the bias is None originally.

File: optimum/fx/optimization/transformations.py [hunk 7]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+
+    preserves_computation = True
+
+    def transform(self, graph_module: "GraphModule") -> "GraphModule":
+        modules = dict(graph_module.named_modules())
+
+        for node in graph_module.graph.nodes:
+            if node.op == "call_module" and node.args[0].op == "call_module":
+                if (
+                    type(modules[node.target]) is torch.nn.BatchNorm2d
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/fx/optimization/transformations.py]: Done in this commit: https://github.com/fxmarty/optimum/commit/9c2b7bd6f358ee9804ea39b8ff4c6513a3dadbb3

File: optimum/fx/optimization/transformations.py [hunk 8]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+    from transformers import ResNetModel, ResNetConfig
+
+    import torch
+    import time
+
+    from transformers.utils.fx import symbolic_trace
+
+    from optimum.fx.optimization import FuseBatchNorm2dInConv2d
+
+    def benchmark(model, inp, iters=20):
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/fx/optimization/transformations.py:484]: We could have very basic examples with minimal `nn.Module` in the docstring, and have a tutorial where we do more involved transforms on real models / combine them. What do you think?

File: optimum/fx/optimization/transformations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/fx/optimization/transformations.py:574]: Can confirm I did not find a way to do it inplace; maybe related to the fact that the bias may be None originally. Do you have an idea? I tried `conv2d.weight.data.copy_(...)`, `conv2d.weight = torch.nn.Parameter(...)` without success. Otherwise I suggest to leave it as is

File: optimum/fx/optimization/transformations.py [hunk 9]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+                    type(modules[node.target]) is torch.nn.BatchNorm2d
+                    and type(modules[node.args[0].target]) is torch.nn.Conv2d
+                ):
+                    if len(node.args[0].users) > 1:  # Output of conv is used by other nodes
+                        continue
+
+                    fused_conv = self.fuse(conv2d=modules[node.args[0].target], bn2d=modules[node.target])
+
+                    # replace the old nn.Conv2d by the fused one
+                    parent_name, _, name = node.args[0].target.rpartition(".")
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/fx/optimization/transformations.py:547]: In this case `parent_name` is `""` and `graph_module.get_submodule(parent_name)` returns the parent module. I added a test for this.

File: optimum/fx/optimization/transformations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/fx/optimization/transformations.py:582]:
```python
conv2d.weight = torch.nn.Parameter(
            conv2d.weight * (bn_w * bn_var_rsqrt).reshape([-1] + [1] * (len(conv2d.weight.shape) - 1))
        )
```
This does not work? What's the issue?
For instance for `MergeLinears`, I am able to do the following:
```python
merged_linear.weight = torch.nn.Parameter(new_weight)
```

File: optimum/fx/optimization/transformations.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/fx/optimization/transformations.py:555]: Alright but let's say you have an attribute called "conv" in your model, at the top of the hierarchy.
In this case `node.target = "conv"`,  `rpartition` would give parent_name = "", and name = "conv"? Just checking that it works, I had a similar issue before.

File: optimum/fx/optimization/transformations.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/fx/optimization/transformations.py:492]: I agree, let's keep the docstring minimal on this, just showing how to apply this to some relevant model, then have a guide / tutorial showing benchmarking stuff.

File: optimum/fx/optimization/transformations.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/fx/optimization/transformations.py:555]: Yes: https://github.com/fxmarty/optimum/blob/7ed5ba7f0d4d9dd1047921b2bcee708e34906b75/tests/fx/test_transformations.py#L414-L436

File: optimum/fx/optimization/transformations.py [hunk 10]
```diff
@@ -463,6 +464,292 @@ def reverse(self, graph_module: "GraphModule") -> "GraphModule":
...
+                        setattr(modules[parent_name], name, fused_linear)
+
+                        # delete batchnorm from the modules
+                        parent_name, _, name = node.target.rpartition(".")
+                        delattr(modules[parent_name], name)
+
+                        node.replace_all_uses_with(node.args[0])
+
+                        graph_module.graph.erase_node(node)  # delete BatchNorm1d
+            elif (
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/fx/optimization/transformations.py]: I removed this transform as it's too custom. I'll probably host it on an other repo.

File: tests/fx/optimization/test_transformations.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fxmarty) [reply tests/fx/optimization/test_transformations.py:222]: @michaelbenayoun I rolled back to having a subfolder. I use `python -m pytest fx/optimization/test_transformations.py --exitfirst` in the yml because `test_quantization.py` is currently not run on main, and the tests don't pass; independenty from this PR (`python -m unittest discover -s fx -p 'test_*.py'` was not capturing it, see e.g. https://github.com/huggingface/optimum/actions/runs/3068109805/jobs/4955165866 )

File: .github/workflows/test_fx.yml [hunk 11]
```diff
@@ -29,9 +29,9 @@ jobs:
...
     - name: Install dependencies
       run: |
-        pip install .
+        pip install .[tests]
         pip install git+https://github.com/huggingface/transformers.git
     - name: Test with unittest
       working-directory: tests
       run: |
-        python -m unittest discover -s fx -p 'test_*.py'
+        python -m pytest fx/optimization/test_transformations.py --exitfirst
```
Comment (michaelbenayoun) [inline .github/workflows/test_fx.yml:37]: No we need to be able to also discover other files, for instance when quantization will be merged, we will need to discover those tests as well.

File: setup.py [hunk 12]
```diff
@@ -22,7 +22,7 @@
     "huggingface_hub>=0.8.0",
 ]
 
-TESTS_REQUIRE = ["pytest", "requests", "parameterized", "pytest-xdist", "Pillow"]
+TESTS_REQUIRE = ["pytest", "requests", "parameterized", "pytest-xdist", "Pillow", "parameterized"]
```
Comment (michaelbenayoun) [inline setup.py:25]: It's already in the list.

File: tests/fx/optimization/test_transformations.py [hunk 13]
```diff
@@ -267,3 +269,105 @@ def test_change_truediv_to_mul_by_inverse():
...
+
+    transformation = FuseBatchNorm1dInLinear()
+    transformed_model = transformation(traced_model)
+
+    num_batchnorm1d = sum(1 if isinstance(mod, torch.nn.BatchNorm1d) else 0 for mod in transformed_model.modules())
+    assert num_batchnorm1d == 0, "there should be no BatchNorm1d left in the model after the transformation"
+
+    output_transformed = transformed_model(**dummy_input)
+
+    assert torch.allclose(output_transformed["logits_per_image"], output_original.logits_per_image, atol=1e-6)
```
Comment (michaelbenayoun) [inline tests/fx/optimization/test_transformations.py:325]: Isn't it already checked in `_check_original_and_transformed_outputs_match`?

**Review by michaelbenayoun — APPROVED** [review]
Left a few comments but you should be able to merge after, thanks!!!

File: tests/fx/optimization/test_transformations.py (hunk 13 above)
Comment (fxmarty) [reply tests/fx/optimization/test_transformations.py:325]: Right, currently `_check_original_and_transformed_outputs_match` handles only NLP tasks, I will modify it to simplify the tests.

**Comment by jogepari** [comment]
Hey, @fxmarty! Found this PR following an old ORT issue to grab your folding implementation, thx!

Sorry for digging an old thread, just wanted to comment that there might be an error [here](https://github.com/huggingface/optimum/blob/7040ff501afa1c60bd5e44af9f828c2ceb6a5ead/optimum/fx/optimization/transformations.py#L551) and [here](https://github.com/huggingface/optimum/blob/7040ff501afa1c60bd5e44af9f828c2ceb6a5ead/optimum/fx/optimization/transformations.py#L673):
`bn_b = bn1d.bias if bn1d.bias is not None else torch.ones_like(bn1d.running_mean)`

In case of affine=False, bias should probably be `torch.zeros_like(bn1d.running_mean)`.

**Comment by fxmarty** [comment]
Hi @jogepari Thank you for notifying, will fix!
